Add integration tests for talks app

boxofrad opened the pull request:
Hi Gophers!

There's a couple of tickets against the talks app I'd like to take a swing at, but thought it made sense to write some tests first as I noticed there weren't any :smile: 

I've tried to change the production code as little as possible / only where necessary to mock out external services.

I submitted my CLA today too...

Have at it! :heart:

boxofrad commented:
This has been sitting around for a while, I understand you're very busy people - would you like to just close this now? :smile_cat:

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+const importPath = "github.com/user/repo/path/to/presentation.slide"
+
+func TestHome(t *testing.T) {
+	i, _ := aetest.NewInstance(nil)
```
Don't ignore errors

campoy commented:
Sorry, working on it now :-)

campoy commented on talksapp/main.go:
```diff
@@ -37,6 +37,10 @@ var (
 	homeArticle       = loadHomeArticle()
 	contactEmail      = "golang-dev@googlegroups.com"
 	githubCredentials = parseGithubCredentials()
```
Can you merge master? This has changed  a bit ago.

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+}
+
+func TestPresentation(t *testing.T) {
+	presentationSrc := `
```
You could use 
```go
presentationSrc := []byte(`
My awesome presentation!
...
`)
```

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+		t.Fatalf("expected status: %d, got: %d", http.StatusOK, w.Code)
+	}
+
+	if !strings.Contains(w.Body.String(), "My awesome presentation") {
```
Should "My awesome presentation" be a variable to factor it out of this and the definition of presentationSrc?

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+	c := appengine.NewContext(r)
+	_, err := memcache.Get(c, importPath)
+
+	if err == memcache.ErrCacheMiss {
```
This should also fail if err != nil

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+- Bar
+- Baz
+`
+	getPresentation = func(client *http.Client, importPath string) (*gosrc.Presentation, error) {
```
set getPresentation to its previous value at the end of this test to avoid side effects

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+}
+
+func TestPresentationNotFound(t *testing.T) {
+	getPresentation = func(client *http.Client, importPath string) (*gosrc.Presentation, error) {
```
defer set back to original value

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+			r.ParseForm()
+
+			if r.Form.Get("version") != version {
+				t.Fatal("expected version to be sent to play.golang.org")
```
log the value of version too

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+			}
+
+			if r.Form.Get("body") != body {
+				t.Fatal("expected body to be sent to play.golang.org")
```
log the value of body too

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+	)
+	defer server.Close()
+
+	playCompileUrl = server.URL
```
defer setting back to original value

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+	i, _ := aetest.NewInstance(nil)
+	defer i.Close()
+
+	form := url.Values{}
```
```go
form := url.Values {
    "version": []string{version},
    "body": []string{body},
}
```

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+	form.Set("body", body)
+
+	r, _ := i.NewRequest("POST", "/compile", nil)
+	r.PostForm = form
```
```go
r.PostForm = url.Values {
    "version": []string{version},
    "body": []string{body},
}
```

boxofrad commented:
@campoy Thanks so much for taking the time to review my PR!

I've rebased from master, fixed the things you commented on and squashed my commits down to keep the commit log nice and clean.

I also added a note in the `README.md` with instruction on running the tests.

Thanks again :gift_heart:

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,293 @@
+const importPath = "github.com/user/repo/path/to/presentation.slide"
+
+func TestHome(t *testing.T) {
+	i, err := aetest.NewInstance(nil)
```
Define an extra function to factor out all the instance and request creation:

```go
func do(t *testing.T, method, path string, f func(*http.Request)) {
        i, err := aetest.NewInstance(nil)
        if err != nil {
                t.Fatal(err)
        }   
        defer i.Close()

        r, err := i.NewRequest(method, path, nil)
        if err != nil {
                t.Fatal(err)
        }   

        f(r)
}
```

Then your code can be as simple as:

```go
func TestHome(t *testing.T) {
        do(t, "GET", "/", func(r *http.Request) {
                w := httptest.NewRecorder()
                handlerFunc(serveRoot).ServeHTTP(w, r)
                if w.Code != http.StatusOK {
                        t.Fatalf("expected status: %d, got: %d", http.StatusOK, w.Code)
                }   

                if !strings.Contains(w.Body.String(), "go-talks.appspot.org") {
                        t.Fatal("expected response to contain: go-talks.appspot.org")
                }   
        })  
}
```

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,293 @@
+	)
+	defer server.Close()
+
+	originalPlayCompileUrl := playCompileUrl
```
You can replace these three lines with:

```go
defer func(old string) { playCompileUrl = old } (playCompileUrl)
playCompileUrl = server.URL
```

campoy commented on talksapp/main.go:
```diff
@@ -36,6 +36,10 @@ var (
+
+	// used for mocking in tests
+	getPresentation = gosrc.GetPresentation
+	playCompileUrl  = "http://play.golang.org/compile"
```
rename to `playCompileURL`

boxofrad commented:
@campoy thanks for the great tips, the tests are looking much cleaner now!

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+			formBody := r.FormValue("body")
+
+			if formVersion != version {
+				t.Fatal("expected version sent to play.golang.org to be: %s, was: %s", version, formVersion)
```
this should be Fatalf

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+			}
+
+			if formBody != body {
+				t.Fatal("expected body sent to play.golang.org to be: %s, was: %s", body, formBody)
```
this should be Fatalf

campoy commented:
Almost there, fix the last two comments and I'll merge.
Thanks for this PR!

boxofrad commented:
Awesome! I've fixed the comments... do you want me to squash my most recent commits?

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+}
+
+func TestPresentationNotFound(t *testing.T) {
+	do(t, "GET", "/"+importPath, func(r *http.Request) {
```
You're missing replacing the `getPresentation` value.

campoy commented on talksapp/main_test.go:
```diff
@@ -0,0 +1,228 @@
+	})
+}
+
+func TestWrongMethod(t *testing.T) {
```
You're missing replacing the getPresentation value.

boxofrad replied:
Do we need to replace it here? if the request method is bad it should never actually hit `getPresentation`

campoy commented:
It seems during the last modif you removed too much code :smile: 

Tests don't pass

boxofrad commented:
Whoops! my bad :flushed: it was 2AM... fixing now :smiley:

boxofrad commented:
Have added in that missing mock... tests are passing on my machine (albeit noisily).
Just to double check, are you running the tests with appengine's tool? `goapp test`

campoy commented:
Works for me, thanks!
